### PR TITLE
[FEATURE] Add Xinference implementation #1179

### PR DESCRIPTION
**Pull request by alvinlee518**
## Issue
Contributes to https://github.com/langchain4j/langchain4j/issues/1179

## Change
Support for [Xorbits Inference](https://inference.readthedocs.io/en/latest/getting_started/index.html).

- Added a XinferenceChatModel
- Added a XinferenceEmbeddingModel
- Added a XinferenceImageModel
- Added a XinferenceLanguageModel
- Added a XinferenceScoringModel
- Added a XinferenceStreamingChatModel
- Added a XinferenceStreamingLanguageModel

## General checklist

- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit tests in all modules, and they are all green
- [x] I have manually run all integration tests in the module I have added/changed, and they are all green

## Checklist for adding new maven module
- [x] I have added my new module in the root pom.xml and langchain4j-community-bom/pom.xml

### CODE REVIEW

**Comment by Martin7-1** [comment]
@alvinlee518 Hi! Thank you for your contribution. Will review it asap.

BTW, it looks like XInference support [docker container](https://github.com/xorbitsai/inference?tab=readme-ov-file#docker). Would it possible to use testcontainers in IT?

**Comment by Martin7-1** [comment]
BTW, could you make all static method call by using `import static`? (e.g. `InternalXinferenceHelper.toTools` -> `toTools`). Just for consistency with other modules.

And, could you please add tests about `ChatModelListener` likt [OllamaChatModelListenerIT](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaChatModelListenerIT.java)

**Comment by alvinlee518** [comment]
> BTW, could you make all static method call by using `import static`? (e.g. `InternalXinferenceHelper.toTools` -> `toTools`). Just for consistency with other modules.
> 
> And, could you please add tests about `ChatModelListener` like [OllamaChatModelListenerIT](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-ollama/src/test/java/dev/langchain4j/model/ollama/OllamaChatModelListenerIT.java)?

Okay, got it.

**Comment by alvinlee518** [comment]
> @alvinlee518 Hi! Thank you for your contribution. Will review it asap.
> 
> BTW, it looks like XInference support [docker container](https://github.com/xorbitsai/inference?tab=readme-ov-file#docker). Would it possible to use testcontainers in IT?

It's a bit tricky for me to test the docker container locally, but I can try adding it.

**Comment by alvinlee518** [comment]
@Martin7-1 The Xinferenrece testcontainers require GPU support. How can I configure it in GitHub Actions to ensure the unit tests pass?

**Comment by Martin7-1** [comment]
Since we're using the free tier, Github Actions do not support GPU... Are there any ways that Xinference can run on CPU (such as `latest-cpu` version). BTW, could we use smaller models (I don't know how much time you test with GPU locally, but if we change it to CPU, it may cost much time, so we need to make the model as small as possible) to reduce IT cost?

If the test takes too long to run with CPU, maybe I'll consider disabling Xinference's IT in Github Actions and just running it locally.

Xinference's ITs are like Ollama's ITs in that they both take a long time to execute tests, and I'm also thinking about and trying to figure out how to optimise them and get them working perfectly in Github Actions. You can try the ideas I suggested above, but if it doesn't work that's fine, I'd run those test cases locally. I think I will review it next week :)

**Comment by alvinlee518** [comment]
@Martin7-1 I have modified the image to the CPU version. When executed in GitHub Actions, the entire process takes approximately 10 minutes.
ImageModel, VisionModel, and tool call streaming functions rely on GPUs and can be validated locally.

**Comment by Martin7-1** [comment]
> @Martin7-1 I have modified the image to the CPU version. When executed in GitHub Actions, the entire process takes approximately 10 minutes. ImageModel, VisionModel, and tool call streaming functions rely on GPUs and can be validated locally.

Thank you! Will try to review it this week.

File: models/langchain4j-community-xinference/src/main/java/dev/langchain4j/community/model/xinference/XinferenceStreamingChatModel.java [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+                log.warn("Exception while calling model listener", e);
+            }
+        });
+        XinferenceStreamingResponseBuilder responseBuilder = new XinferenceStreamingResponseBuilder();
+        client.chatCompletions(request)
+                .onPartialResponse(partialResponse -> {
+                    responseBuilder.append(partialResponse);
+                    final List<ChatCompletionChoice> choices = partialResponse.getChoices();
+                    if (!isNullOrEmpty(choices)) {
+                        for (final ChatCompletionChoice choice : choices) {
```
Comment (Martin7-1) [inline models/langchain4j-community-xinference/src/main/java/dev/langchain4j/community/model/xinference/XinferenceStreamingChatModel.java:177]: I'm not sure about this `for-each`, if `n` is not equal to 1, which means `handler.onNext` should handle different choice. I think it's difficult and is not supported in `langchain4j` now. Let's only consider `n=1`. Please refer to [langchain4j-open-ai](https://github.com/langchain4j/langchain4j/blob/20c6f3c73adbfc8698850050822437d6865888a2/langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiStreamingChatModel.java#L273). WDYT?

File: models/langchain4j-community-xinference/src/main/java/dev/langchain4j/community/model/xinference/XinferenceStreamingLanguageModel.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+import static dev.langchain4j.spi.ServiceHelper.loadFactories;
+
+public class XinferenceStreamingLanguageModel implements StreamingLanguageModel {
+
+    private final XinferenceClient client;
+    private final String modelName;
+    private final Integer maxTokens;
+    private final Double temperature;
+    private final Double topP;
+    private final Integer n;
```
Comment (Martin7-1) [inline models/langchain4j-community-xinference/src/main/java/dev/langchain4j/community/model/xinference/XinferenceStreamingLanguageModel.java:30]: Same as `XinferenceChatModel`

File: models/langchain4j-community-xinference/src/main/java/dev/langchain4j/community/model/xinference/XinferenceStreamingLanguageModel.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                .stop(stop)
+                .presencePenalty(presencePenalty)
+                .frequencyPenalty(frequencyPenalty)
+                .user(user)
+                .build();
+        final XinferenceStreamingResponseBuilder responseBuilder = new XinferenceStreamingResponseBuilder();
+        client.completions(request)
+                .onPartialResponse(completionResponse -> {
+                    responseBuilder.append(completionResponse);
+                    for (final CompletionChoice choice : completionResponse.getChoices()) {
```
Comment (Martin7-1) [inline models/langchain4j-community-xinference/src/main/java/dev/langchain4j/community/model/xinference/XinferenceStreamingLanguageModel.java:106]: Same as `XinferenceStreamingChatModel`

**Review by Martin7-1 — COMMENTED** [review]
@alvinlee518 Thank you!

**Review by Martin7-1 — COMMENTED** [review]
@alvinlee518 Thank you! About IT: maybe we should give up comitting new image, because so much images will exceed Github Actions disk space limit. WDYT?

**Comment by alvinlee518** [comment]
> @alvinlee518 Thank you! About IT: maybe we should give up comitting new image, because so much images will exceed Github Actions disk space limit. WDYT?
@Martin7-1 
The new image is intended to avoid re-downloading the model. If you're concerned about the disk space limit, I can remove it.

File: models/langchain4j-community-xinference/src/main/java/dev/langchain4j/community/model/xinference/InternalXinferenceHelper.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        }
+        return messages.stream()
+                .map(msg -> {
+                    if (msg instanceof dev.langchain4j.data.message.SystemMessage message) {
+                        return SystemMessage.of(message.text());
+                    } else if (msg instanceof AiMessage message) {
+                        if (!message.hasToolExecutionRequests()) {
+                            return AssistantMessage.of(message.text());
+                        }
+                        final List<ToolCall> list = message.toolExecutionRequests().stream()
```
[the remark this answers is not in the document]
Comment (alvinlee518) [reply models/langchain4j-community-xinference/src/main/java/dev/langchain4j/community/model/xinference/InternalXinferenceHelper.java:60]: It's just to clarify that it is a read-only reference.

**Review by Martin7-1 — APPROVED** [review]
@alvinlee518 Thank you! LGTM. Could you please add docs in langchain4j main repo?

**Comment by alvinlee518** [comment]
> @alvinlee518 Thank you! LGTM. Could you please add docs in langchain4j main repo?

Okay, I'll add it later.

**Pull request merged**